cmd/nv2/gencert.go

```diff
@@ -0,0 +1,184 @@
+package main
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"log"
+	"math/big"
+	"net"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/urfave/cli/v2"
+)
+
+const (
+	KEYS_BASE_DIR = ".notary"
+)
+
+var generateCertCommand = &cli.Command{
+	Name:      "generate-certificates",
+	Usage:     "Generates a test crt and key file.",
+	ArgsUsage: "[host]",
+	Flags: []cli.Flag{
+		&cli.IntFlag{
+			Name:     "rsaBits",
+			Usage:    "rsaBits 2048",
+			Required: false,
+			Value:    2048,
+		},
+	},
+	Action: runGenerateCert,
+}
+
+//ref: https://golang.org/src/crypto/tls/generate_cert.go
+func publicKey(priv interface{}) interface{} {
+	switch k := priv.(type) {
+	case *rsa.PrivateKey:
+		return &k.PublicKey
+	default:
+		return nil
+	}
+}
+
+func ensureKeysDir() (string, error) {
+	dirname, err := os.UserHomeDir()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	keysDir := filepath.Join(dirname, KEYS_BASE_DIR, "keys")
+	_, err = os.Stat(keysDir)
+	if os.IsNotExist(err) {
+		err = os.MkdirAll(keysDir, 0700) // Only user has permissions on this directory
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return keysDir, nil
+}
+
+func runGenerateCert(ctx *cli.Context) error {
+
+	host := ctx.Args().First()
+	if len(host) == 0 {
+		return errors.New("Missing required [host] parameter")
+	}
+
+	rsaBits := ctx.Int("rsaBits")
+	var priv interface{}
+	var err error
+
+	fmt.Printf("Generating RSA Key with %d bits\n", rsaBits)
+	priv, err = rsa.GenerateKey(rand.Reader, rsaBits)
+
+	if err != nil {
+		return fmt.Errorf("Failed to generate private key: %v", err)
+	}
+
+	// ECDSA, ED25519 and RSA subject keys should have the DigitalSignature
+	// KeyUsage bits set in the x509.Certificate template
+	keyUsage := x509.KeyUsageDigitalSignature
+	// Only RSA subject keys should have the KeyEncipherment KeyUsage bits set. In
+	// the context of TLS this KeyUsage is particular to RSA key exchange andgit st
```

Comment:
I'm not sure why keyencipherment is set. Are these certs used for TLS somehow?

Reply:
I'll remove all TLS related usage here since these are only tests certs for signing.